components_library/notes.md

```diff
@@ -0,0 +1,62 @@
+# Notes setting up RISC-V FS in components
+
+## Thoughts
+
+I'm going to try to not use SimpleBoard and see what happens.
+...
+Nope, that was a bad idea.
+SimpleBoard's `connectThings` is useful and so it the Clock setup in the constructor.
+
+Setting up the I/O stuff is pretty awkward.
+For RISC-V it's going to be spread between the `RiscvBoard` and the `HiFive` platform.
+
+Also, setting up the I/O stuff is quite strange.
+I'm not sure what should be done by the board and what is done in the caches.
+
+## Things to do
+
+- Figure out a better solution for the dtb generation
+  - Make is so that we don't have to use the variable name "disk" for the platform device
+- Check if it works with no_cache. It should as the fs_linux.py file seems to.
+
+- MMU problems
+  - Need to have MMU caches when there are caches
+  - Or, if there are no caches, then need to have another bus to connect MMUs
+
+## General components improvements
+
+- Add a "CompatibilityError" instead of "EnvironmentError"
+- make `is_ruby` a member of the cache hierarchy, not an "environment" variable.
+- Add a helper for disk images?
+- Limitations of SimpleBoard doesn't seem true anymore
+- When can I use the `get_processor` and `get_cores` functions?
+- In simple_core can't import all CPU types (e.g., KVM CPU) as they may not be built
+- membus should be a class member, and we shouldn't set the badaddr responder every time _get_default_membus is called
+- I don't love the way the createInterrupts() functions are called. This should probably be either in the `__init__` function or in a different function than `connect_interrupts`. I think the board should be responsible for this, not the caches.
+
+## Questions about components
+
+- What's the point of SimpleBoard and why does X86Board inherit it?
+  - Has some useful functions... document this
+- Where do I get the memory size? Is this a parameter memory that the system should grab, or is it a parameter that system should set on the memory?
+  - Needs to be fixed
+- In the x86 board, which way does the memory go? Do we set the memory range on the board from the memory device or vice versa?
+
+## Questions about RISC-V
+
+- Should we use the bare metal workload or `RiscvLinux`?
+  - The difference is the dtb file baremetal assumes the dtb file is compiled into the bootloader. The `RiscvLinux` workload can load a the dts file from disk.
+- How much of HiFive platform could be moved out into the components?
+- How does virtio disk image work?
+- The PMA checker shouldn't have an empty default in the RiscvMMU. This is a recipe for disaster
+
+## Things that could be useful
+
+A `require` function which, for instance, requires that ISA is RISCV.
```

Comment:
Yes, I think a more clever way to delcare run-time requirements would be useful and remove a lot of boilerplate. For example, when I write a config script I'm having to put a lot of if-statements at the top making sure everything is ok, otherwise return an error (e.g., "You need to have gem5 compiled to X86").